tidy3d/packaging.py

```diff
@@ -0,0 +1,148 @@
+"""
+This file contains a set of functions relating to packaging tidy3d for distribution. Sections of the codebase should depend on this file, but this file should not depend on any other part of the codebase.
+
+This section should only depend on the standard core installation in the pyproject.toml, and should not depend on any other part of the codebase optional imports.
+"""
+import functools
+import numpy as np
+from typing import Literal
+
+from .exceptions import Tidy3dImportError
+
+vtk = {
+    "mod": None,
+    "id_type": np.int64,
+    "vtk_to_numpy": None,
+    "numpy_to_vtkIdTypeArray": None,
+    "numpy_to_vtk": None,
+}
+
+
+def check_import(module_name: str) -> bool:
+    """
+    Check if a module or submodule section has been imported. This is a functional way of loading packages that will still load the corresponding module into the total space.
+
+    Parameters
+    ----------
+    module_name
+
+    Returns
+    -------
+    bool
+        True if the module has been imported, False otherwise.
+
+    """
+    try:
+        __import__(module_name)
+        return True
+    except ImportError:
+        return False
+
+
+def verify_packages_import(modules: list, required: Literal["either", "all"] = "all"):
+    def decorator(func):
+        """
+        When decorating a method, requires that the specified modules are available. It will raise an error if the
+        module is not available depending on the value of the 'required' parameter which represents the type of
+        import required.
+
+        There are a few options to choose for the 'required' parameter:
+        - 'all': All the modules must be available for the operation to continue without raising an error
+        - 'either': At least one of the modules must be available for the operation to continue without raising an error
+
+        Parameters
+        ----------
+        func
+            The function to decorate.
+
+        Returns
+        -------
+        checks_modules_import
+            The decorated function.
+
+        """
+
+        def checks_modules_import(*args, **kwargs):
+            """
+            Checks if the modules are available. If they are not available, it will raise an error depending on the value.
+            """
+            available_modules_status = []
+            maximum_amount_modules = len(modules)
+
+            module_id_i = 0
```

Comment:
This loop seems to try to import all modules even is "any"/"either" is required. I think I'd invert the logic here. Something like:
```python
if required == "all":
    if not all(check_import(m) for m in modules):
        raise ...
elif required == "any"
    if not any(check_import(m) for m in modules):
        raise ...
else # invalid 'required' value
    raise ...
```

Reply:
Ahh yeah this is more operationally efficient! Implementing.

Thanks for all the feedback